packages/knit-gateway/src/stitch.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright 2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { Schema_Field } from "@buf/bufbuild_knit.bufbuild_es/buf/knit/gateway/v1alpha1/knit_pb.js";
+import type {
+  JsonObject,
+  AnyMessage,
+  PlainMessage,
+  IMessageTypeRegistry,
+} from "@bufbuild/protobuf";
+import type { Relation } from "./gateway.js";
+import { formatValue, type Patch } from "./json.js";
+import { Code, ConnectError } from "@bufbuild/connect";
+
+interface Batch {
+  bases: AnyMessage[];
+  formatTargets: JsonObject[];
+  field: PlainMessage<Schema_Field> & { relation: Relation };
+}
+
+/**
+ * Stitches the patches into the targets.
+ */
+export async function stitch(
+  patches: Patch[],
+  typeRegistry?: IMessageTypeRegistry
+) {
+  while (patches.length > 0) {
+    const batches = makeBatches(patches);
+    const resolves: Promise<Patch[][]>[] = [];
+    for (const batch of batches) {
+      resolves.push(resolveBatch(batch, typeRegistry));
+    }
+    patches = (await Promise.all(resolves)).flat(2);
```

Comment:
If `resolveBatch` returned `Patch[]`, you wouldn't need the call to `.flat(2)` here, right?

Reply:
We'd still need to call `.flat()` but updated the code!